### DISCUSSION TITLE
dmass correction

### DISCUSSION THREAD

**Discussion by joe-pitt** [uataq/stilt#131] [Users]
Hi Ben,

Excuse my ignorance here, but I've been comparing the wrappers from the original version of STILT (i.e. http://stilt-model.org) to yours, and I wondered if you could explain one of the differences for me. In the original wrappers there is an option to correct for violation of mass conservation in the met fields (dmassTF in Trajecfoot.r), which defaults to TRUE (at least in my copy). I couldn't find an equivalent option in your wrappers - have I missed it or did you decide not to implement this correction?

Cheers,
Joe

**Comment by benfasoli** [comment]
Hi Joe, I just chatted with John Lin about his view on the dmass correction. He suggested -

> The dmass correction is a first order correction that may not work when the mass violation is too large.  I suggest that the dmass correction still be left in so that the option to carry out the correction is still there, but the default be set to FALSE instead so that people need to be more aware and think about the fact that they are setting the dmass correction to TRUE when they are running STILT.  The users can also compare dmass = TRUE versus FALSE to get a sense of the impact of mass violation.

So we haven't implemented because we aren't actively using this flag but I'd be happy to add it. For now, you could use existing methods to replicate the behavior. Make sure `dmas` is included in the `varsiwant` vector and pass the following to the `before_footprint` argument -

```r
before_footprint <- function() {
  mean_dmas_by_time <- output$particle %>% group_by(time) %>% summarize(mean_dmas = mean(dmas))
  output$particle <- left_join(output$particle, mean_dmas_by_time, by = 'time')
  output$particle$dmas <- output$particle$dmas / output$particle$mean_dmas
  output$particle$foot <- output$particle$foot * output$particle$dmas
  output
}
```

This mutates the `output` object which contains the trajectory information to normalize the `dmas` value per timestep and rescale the footprint sensitivity associated with each particle prior to gridding the footprint.

**Comment by joe-pitt** [comment]
Hi Ben, Thanks for looking into this so quickly! That's really useful to know - it sounds like it would be sensible for me to repeat some of my runs with the dmass correction off to make sure I don't get a radically different result.  And I hadn't noticed the before_trajec and before_footprint options - that's a neat way to implement things like this!

Cheers,
Joe

**Comment by benfasoli** [comment]
Hey Joe, just circling back to this - did you do any sensitivity tests for the dmass correction?

**Comment by joe-pitt** [comment]
Hi Ben - thanks for the reminder! I've been clogging up the queues on the HPC this past month so hadn't got round to this, but I've just set some repeat runs going with dmassTF=FALSE. I'll get back to you once I have some results...
